### DISCUSSION TITLE
[Documentation] Windows avaliablity

### DISCUSSION THREAD

**Discussion by syrull** [raiyanyahya/freshenv#31] [Future]
Maybe it is good to note that it is not supported for windows, do you plan to release that for Windows? If so I would like to contribute to that the idea is awesome!

**Comment by raiyanyahya** [comment]
@syrull thanks. I will add it to the readme. I do plan to work on windows support after working on the current backlog :
1. Custom Environments.
2. Cloud Share and Save.

I will let you know when I start working on Windows support. Hope this helps 😀

**Reply by syrull** [reply]
Awesome, do you have any particular replacements for `dockerpty`?

**Reply by raiyanyahya** [reply]
Nope. Unfortunetlybwindows does not support the functl module. Will have to find other alternatives or build something myself.

**Reply by syrull** [reply]
I've explored some solutions, unfortunately, the tty (or pty) is not avaliable in Windows at all. As per https://github.com/docker/docker-py/issues/390 it seems really hard/messy to do so. And in my short research and a little bit of hacking the best working solution is `os.system` and passing the arguments in a raw format. 

Note:
For some reason `subprocess.Popen` is really really slow and not interactive (didn't really bother to see why).
 
I believe that it is possible to create a context manager with that, I have to research more, and possibly I can create a PoC on a fork of the repository if that's fine by you :) @raiyanyahya

**Comment by raiyanyahya** [comment]
:) I have tried with subprocess, its really tricky and you cant really guarantee long running command outputs. It doesnt come close to the tty experience. You dont want different experience based on different OS. Also, windows doesnt run docker native and require hyper-v or some other form of virtualization. You also need docker desktop. I am however testing this on cygwin and wsl for windows. I feel that would be the initial support on Windows. 

I would request you to wait on the POC fork since I am investing all my time on custom environments and cloud features and I want to finish it by June :)
I will let you know when I start looking into native windows support. 
Thank you for the interest, keeps me motivated.
Cheers!

**Reply by syrull** [reply]
Yep, I agree with your points, docker indeed runs a bit tricky on windows. The support in WSL is currently working (tested on Ubuntu 20.04 WSL2) and that is expected. The docker should be exposed to the wsl2 distro beforehand though.

However, the problems are remaining the same with Cygwin. I don't think that this is going to work without any major changes to the source code natively on windows. 

One thing that I have in mind is how the workspaces work in VSCode (via the `Remote-Containers` extension), maybe you can drive some inspiration from there because essentially I think that it does a very similar thing to your project. 

Also we are just discussing some stuff, I completely understand that you have other things in mind but keeping this convo for further references while working on the port to Windows (evnetually). And always I am open to exploring different ideas and stuff like that!
